crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+    dropped: bool,
+}
+
+// Need this because SignalKind does not have default
+impl Default for JoinHandle {
+    fn default() -> Self {
+        JoinHandle {
+            task: OnceCell::new(),
+            allowed_signal: SignalKind::terminate(),
+            dropped: false,
+        }
+    }
+}
+
+impl JoinHandle {
+    /// Creates a new join handle from a task.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use jstzd::task::task::JoinHandle;
+    /// use tokio::signal::unix::SignalKind;
+    ///
+    /// # tokio_test::block_on(async {
+    /// let handle = JoinHandle::new(
+    ///     tokio::spawn(async move {
+    ///         tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    ///     }),
+    ///     SignalKind::terminate(),
+    /// );
+    /// # });
+    /// ```
+    pub fn new(task: tokioJoinHandle<()>, signal: SignalKind) -> Self {
+        JoinHandle {
+            task: OnceCell::from(task),
+            allowed_signal: signal,
+            ..Default::default()
+        }
+    }
+
+    /// This function will block until the task has exited, returning a success or an error
+    async fn wait(&mut self) -> anyhow::Result<()> {
+        return match self.task.take() {
+            Some(handle) => {
+                let result = handle.await;
+                if result.is_err() {
+                    let err = result.unwrap_err();
+                    if err.is_cancelled() {
+                        return anyhow::Ok(());
+                    }
+                    return Err(err.into());
+                }
+                anyhow::Ok(())
+            }
+            None => anyhow::Ok(()),
+        };
+    }
+
+    /// This function will signal the running task with the specified signal
+    /// This is a non-blocking operation
+    fn signal(&self, signal: SignalKind) -> anyhow::Result<()> {
+        if signal != self.allowed_signal && signal != SignalKind::terminate() {
+            return Err(anyhow::anyhow!(
+                "unacceptable signal '{:?}'",
+                signal.as_raw_value()
+            ));
+        }
+        if let Some(handle) = self.task.get() {
+            handle.abort();
+        }
+        return anyhow::Ok(());
+    }
+
+    async fn shutdown(&mut self) {
+        let _ = self.signal(SignalKind::terminate());
+        let _ = self.wait().await;
+    }
+}
+
+impl Future for JoinHandle {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        return match self.task.get() {
+            Some(handle) => {
+                if handle.is_finished() {
+                    Poll::Ready(())
+                } else {
+                    let waker = cx.waker().clone();
+                    tokio::spawn(async move {
+                        tokio::time::sleep(std::time::Duration::from_secs_f32(0.3)).await;
+                        waker.wake();
+                    });
+                    Poll::Pending
```

Comment:
Instead of creating a new thread to wakeup the waker, recommend using the `.poll` method of the `handle` (it also implements `Future`). Something like the following should work:
```rust
  let pin = Pin::new(handle);
  match pin.poll(cx) {
      Poll::Ready(_) => Poll::Ready(())
      Poll::Pending => Poll::Pending,
  }
  ```

Reply:
As discussed offline, calling poll here requires a mutable reference to `handle` and that makes things quite complicated here. I made a few attempts but still could not make it work.